Prevent unmanaged pods in GKE's containerd flavors

bmcustodio opened the pull request:
The changes that we have been doing to `/etc/defaults/kubelet` are reset on node reboots, as is apparently the whole `/etc` directory — which also means that `/etc/cni/net.d/05-cilium.conf` is removed.

This would not be a problem if the assumption we made that the node taint we recommend placing on the nodes would come back upon reboots held true, but in practice it doesn't.

Besides this, it seems that containerd will re-instante its CNI configuration file, and it will do so way before Cilium has had the chance to re-run on the node and re-create its CNI configuration, causing pods to be assigned IPs by the default CNI rather than by Cilium in the meantime.

This commit attempts at preventing that from happening by observing that `/home/kubernetes/bin/kubelet` (i.e. the actual kubelet binary) is kept between reboots and executed concurrently with containerd by systemd. We leverage on this empirical observation to replace this file kubelet with a wrapper script that, under the required conditions, disables containerd, patches its configuration, removes undesired CNI configuration files, re-enables containerd and becomes the kubelet.

Additionally, to prevent situations in which the GKE node is forcibly stopped and re-created from causing unmanaged pods, and building on the observation that the node comes back with the same name and pods are already scheduled there, we change the recommended taint effect from `NoSchedule` to `NoExecute`, to cause any previously scheduled pods to be evicted, preventing them from getting IPs assigned by the default CNI.This should not impact other environments due to the nature of `NoExecute`, so we recommend it everywhere.

```release-note
Prevent unmanaged pods in GKE's containerd flavors.
*Important:* Users should update their node taints from `node.cilium.io/agent-not-ready=true:NoSchedule` to `node.cilium.io/agent-not-ready=true:NoExecute`.
*Important:* During the first node reboot after the fix is applied pods may still get IPs from the default CNI as cilium-node-init is only run later in the node startup process. The fix will then be in place for all subsequent reboots.
```

kaworu commented on install/kubernetes/cilium/files/nodeinit/prestop.bash:
```diff
@@ -30,9 +30,18 @@ rm -f /tmp/node-init.cilium.io
-echo "Changing kubelet configuration to --network-plugin=kubenet"
-sed -i "s:--network-plugin=cni\ --cni-bin-dir={{ .Values.cni.binPath }}:--network-plugin=kubenet:g" /etc/default/kubelet
-echo "Restarting kubelet..."
+if [[ command -v gke && command -v containerd ]];
```
Bash's `[[` will complain here, and I think `then` cannot be omitted:

```suggestion
if command -v gke && command -v containerd; then
```

Question: shouldn't this test look like the one in `startup.bash`?

bmcustodio replied:
That's correct — thank you for spotting it. 🦅👀

kaworu commented on install/kubernetes/cilium/files/nodeinit/startup.bash:
```diff
@@ -22,11 +22,86 @@ fi
+# Check if we're running on a GKE containerd flavor.
+GKE_KUBERNETES_BIN_DIR="/home/kubernetes/bin"
+if [[ -f "${GKE_KUBERNETES_BIN_DIR}/gke" ]] && command -v containerd >/dev/null 2>&1;
+then
```
nit:
```suggestion
if [[ -f "${GKE_KUBERNETES_BIN_DIR}/gke" ]] && command -v containerd &>/dev/null; then
```

kaworu reviewed (changes requested):
Couple of question, but overall LGTM.

kaworu reviewed (approved):
:rocket:

christarazi commented on install/kubernetes/cilium/files/nodeinit/startup.bash:
```diff
@@ -22,11 +22,85 @@ fi
+  # configuration are removed. This leaves room for containerd and its CNI to
+  # take over pods previously managed by Cilium, causing these pods to become
+  # unmanaged. We rely on the fact that /home/kubernetes/bin/kubelet is not
+  # changed across node reboots, and replace it with a wrapper script that
```
```suggestion
  # configuration are removed. This leaves room for containerd and its CNI to
  # take over pods previously managed by Cilium, causing Cilium to lose ownership over these pods. 
  # We rely on the fact that /home/kubernetes/bin/kubelet is not
  # changed across node reboots, and replace it with a wrapper script that
```

christarazi commented on install/kubernetes/cilium/files/nodeinit/startup.bash:
```diff
@@ -22,11 +22,85 @@ fi
+  # the /etc directory and our changes to the kubelet and Cilium's CNI
+  # configuration are removed. This leaves room for containerd and its CNI to
+  # take over pods previously managed by Cilium, causing these pods to become
+  # unmanaged. We rely on the fact that /home/kubernetes/bin/kubelet is not
```
What is the guarantee that the kubelet won't change? Where is that "fact"? I think these are important to document in case this changes in the future.

bmcustodio replied:
Unfortunately I couldn't find any documentation mentioning this fact — this is strictly an "observation", not a documented fact. It's one of the weakness of this fix/workaround, but I guess we'll have to deal with that. I'll update the comment though to change `s/fact/empirical observation/`. :+1:

pchaigno replied:
How will we notice if that assumption breaks? I'm guessing none of our tests will detect that because we don't restart nodes in CI?

bmcustodio replied:
This is of course a concern and a shortcoming of this approach. Unfortunately I don't have a good answer to it other than having CI restart nodes. 😞

christarazi commented on install/kubernetes/cilium/files/nodeinit/startup.bash:
```diff
@@ -22,11 +22,85 @@ fi
+# Only stop and start containerd if the Cilium CNI configuration does not exist,
+# or if the 'conf_template' property is present in the containerd config file,
+# in order to avoid unnecessarily restarting containerd.
+if [[ "$(find "${CNI_CONF_DIR}" -type f -name '*cilium*' | wc -l)" == "0" || \
```
Nit: the less we pipe, the less chance something goes wrong

```suggestion
if [[ -z "$(find "${CNI_CONF_DIR}" -type f -name '*cilium*')" || \
```

christarazi commented on install/kubernetes/cilium/files/nodeinit/startup.bash:
```diff
@@ -22,11 +22,85 @@ fi
+exec /home/kubernetes/bin/the-kubelet "${@}" --network-plugin=cni --cni-bin-dir={{ .Values.cni.binPath }}
+EOF
+  else
+    echo "Kubelet wrapper seems to already exist, skipping..."
```
Nit

```suggestion
    echo "Kubelet wrapper already exists, skipping..."
```

christarazi reviewed (commented):
Just a few minor comments, great work! :rocket:

sayboras commented on Documentation/gettingstarted/k8s-install-default.rst:
```diff
@@ -45,9 +45,9 @@ to create a Kubernetes cluster locally or using a managed Kubernetes service:
            # Create the node pool with the following taint to guarantee that
-           # Pods are only scheduled in the node when Cilium is ready.
+           # Pods are only scheduled/executed in the node when Cilium is ready.
            gcloud container clusters create "${NAME}" \
-            --node-taints node.cilium.io/agent-not-ready=true:NoSchedule \
+            --node-taints node.cilium.io/agent-not-ready=true:NoExecute \
```
based on my understanding, NoExecute and NoShedule is not overlapping, do you think we need to have both ? 

- NoSchedule is to make sure no new pod created util cilium is ready
- NoExecute is to make sure that existing pods get evicted.

bmcustodio replied:
That's a great question — @aanm asked the same. Based on the [documentation](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/), I believe they're indeed overlapping:

> (...) if there is at least one un-ignored taint with effect `NoExecute` then the pod will be evicted from the node (if it is already running on the node), *and* will not be scheduled onto the node (if it is not yet running on the node).

So I think we're good with `NoExecute` — at least that's my reading of the above.

sayboras replied:
thanks, I might have missed such convo :100:

sayboras reviewed (commented):
I have only one comment as per below, this is super findings.

aanm reviewed (changes requested):
I've re-read the documentation about this NoExecute vs NoSchedule. @bmcustodio can we change the taint in our conformance tests (see `.github/workflows/conformance*`) from  `NoSchedule` to `NoExecute`? I'm curious to know if we break anything on our e2e tests before recommending this to users. Thanks!

sayboras commented on Documentation/gettingstarted/requirements-eks.rst:
```diff
@@ -37,11 +37,11 @@ For more information on AWS ENI mode, see :ref:`ipam_eni`.
            - key: "node.cilium.io/agent-not-ready"
              value: "true"
-             effect: "NoSchedule"
+             effect: "NoExecute"
```
nit: NoSchedule -> NoExecute at line 26 in this file as well :)

bmcustodio replied:
Nice catch, thank you!

bmcustodio committed:
Prevent unmanaged pods in GKE's containerd flavors

The changes that we have been doing to /etc/defaults/kubelet are reset
on node reboots, as is apparently the whole /etc directory --- which
also means that /etc/cni/net.d/05-cilium.conf is removed.

This would not be a problem if the assumption we made that the node taint we
recommend placing on the nodes would come back upon reboots held true, but in
practice it doesn't.

Besides this, it seems that containerd will re-instante its CNI
configuration file, and it will do so way before Cilium has had the
chance to re-run on the node and re-create its CNI configuration,
causing pods to be assigned IPs by the default CNI rather than by Cilium
in the meantime.

This commit attempts at preventing that from happening by observing that
/home/kubernetes/bin/kubelet (i.e. the actual kubelet binary) is kept between
reboots and executed concurrently with containerd by systemd. We leverage on
this empirical observation to replace this file kubelet with a wrapper script
that, under the required conditions, disables containerd, patches its
configuration, removes undesired CNI configuration files, re-enables
containerd and becomes the kubelet.

Signed-off-by: Bruno Miguel Custódio <brunomcustodio@gmail.com>
Co-authored-by: Alexandre Perrin <alex@kaworu.ch>
Co-authored-by: Chris Tarazi <chris@isovalent.com>

bmcustodio committed:
Recommend 'NoExecute' instead of 'NoSchedule'.

To prevent situations in which the GKE node is forcibly stopped and
re-created from causing unmanaged pods, and building on the observation
that the node comes back with the same name and pods are already
scheduled there, we change the recommended taint effect from NoSchedule
to NoExecute, to cause any previously scheduled pods to be evicted,
preventing them from getting IPs assigned by the default CNI. This
should not impact other environments due to the nature of 'NoExecute',
so we recommend it everywhere.

Signed-off-by: Bruno Miguel Custódio <brunomcustodio@gmail.com>
Co-authored-by: Tam Mach <sayboras@yahoo.com>

pchaigno reviewed (approved):
I reviewed only my codeowners.

I do have one question at https://github.com/cilium/cilium/pull/18486#discussion_r794057613, to echo Chris's concern.
